File: packages/ios-client/Sources/FishjamClient/webrtc/StatsCollector.swift
import WebRTC

class StatsCollector {
    static func getStats(for connection: RTCPeerConnection) async -> [String: RTCStats] {
        let report = await connection.statistics()
        return extractRelevantStats(rp: report)
    }

    private static func extractRelevantStats(rp: RTCStatisticsReport) -> [String: RTCStats] {
        return rp.statistics.reduce(into: [:]) { stats, pair in
            let stat = pair.value
            switch stat.type {
            case "outbound-rtp": stats[stat.id as String] = extractOutboundStats(from: stat)
            case "inbound-rtp": stats[stat.id as String] = extractInboundStats(from: stat)
            default: break
            }
        }
    }

    private static func extractOutboundStats(from stats: RTCStatistics) -> RTCOutboundStats {
        let duration = stats.values["qualityLimitationDurations"] as? [String: Double]
        let qualityLimitation = QualityLimitationDurations(
            bandwidth: duration?["bandwidth"] ?? 0.0,
            cpu: duration?["cpu"] ?? 0.0,
            none: duration?["none"] ?? 0.0,
            other: duration?["other"] ?? 0.0
        )

        return RTCOutboundStats(
            kind: stats.values["kind"] as? String ?? "",
            rid: stats.values["rid"] as? String ?? "",
            bytesSent: stats.values["bytesSent"] as? UInt ?? 0,
            targetBitrate: stats.values["targetBitrate"] as? Double ?? 0.0,
            packetsSent: stats.values["packetsSent"] as? UInt ?? 0,
            framesEncoded: stats.values["framesEncoded"] as? UInt ?? 0,
            framesPerSecond: stats.values["framesPerSecond"] as? Double ?? 0.0,
            frameWidth: stats.values["frameWidth"] as? UInt ?? 0,
            frameHeight: stats.values["frameHeight"] as? UInt ?? 0,
            qualityLimitationDurations: qualityLimitation
        )
    }

    private static func extractInboundStats(from stats: RTCStatistics) -> RTCInboundStats {
        return RTCInboundStats(
            kind: stats.values["kind"] as? String ?? "",
            jitter: stats.values["jitter"] as? Double ?? 0.0,
            packetsLost: stats.values["packetsLost"] as? UInt ?? 0,
            packetsReceived: stats.values["packetsReceived"] as? UInt ?? 0,
            bytesReceived: stats.values["bytesReceived"] as? UInt ?? 0,
            framesReceived: stats.values["framesReceived"] as? UInt ?? 0,
            frameWidth: stats.values["frameWidth"] as? UInt ?? 0,
            frameHeight: stats.values["frameHeight"] as? UInt ?? 0,
            framesPerSecond: stats.values["framesPerSecond"] as? Double ?? 0.0,
            framesDropped: stats.values["framesDropped"] as? UInt ?? 0
        )
    }
}

File: packages/android-client/FishjamClient/src/main/java/com/fishjamcloud/client/webrtc/StatsCollector.kt
package com.fishjamcloud.client.webrtc

import com.fishjamcloud.client.models.QualityLimitationDurations
import com.fishjamcloud.client.models.RTCInboundStats
import com.fishjamcloud.client.models.RTCOutboundStats
import com.fishjamcloud.client.models.RTCStats
import org.webrtc.PeerConnection
import org.webrtc.RTCStatsReport
import java.math.BigInteger
import kotlin.coroutines.resume
import kotlin.coroutines.suspendCoroutine

class StatsCollector {
  companion object {
    suspend fun getStats(connection: PeerConnection): Map<String, RTCStats> =
      suspendCoroutine { continuation ->
        connection.getStats { report ->
          continuation.resume(extractRelevantStats(report))
        }
      }

    private fun extractRelevantStats(report: RTCStatsReport): Map<String, RTCStats> =
      report.statsMap.values
        .mapNotNull { stats ->
          when (stats.type) {
            "outbound-rtp" -> stats.id to extractOutboundStats(stats)
            "inbound-rtp" -> stats.id to extractInboundStats(stats)
            else -> null
          }
        }.toMap()

    private fun extractOutboundStats(stats: org.webrtc.RTCStats): RTCOutboundStats {
      val durations = stats.members["qualityLimitationDurations"] as? Map<*, *>
      val qualityLimitation =
        QualityLimitationDurations(
          durations?.get("bandwidth") as? Double ?: 0.0,
          durations?.get("cpu") as? Double ?: 0.0,
          durations?.get("none") as? Double ?: 0.0,
          durations?.get("other") as? Double ?: 0.0
        )

      return RTCOutboundStats(
        kind = stats.members["kind"] as? String,
        rid = stats.members["rid"] as? String,
        bytesSent = stats.members["bytesSent"] as? BigInteger,
        targetBitrate = stats.members["targetBitrate"] as? Double,
        packetsSent = stats.members["packetsSent"] as? Long,
        framesEncoded = stats.members["framesEncoded"] as? Long,
        framesPerSecond = stats.members["framesPerSecond"] as? Double,
        frameWidth = stats.members["frameWidth"] as? Long,
        frameHeight = stats.members["frameHeight"] as? Long,
        qualityLimitationDurations = qualityLimitation
      )
    }

    private fun extractInboundStats(stats: org.webrtc.RTCStats): RTCInboundStats =
      RTCInboundStats(
        kind = stats.members["kind"] as? String,
        jitter = stats.members["jitter"] as? Double,
        packetsLost = stats.members["packetsLost"] as? Int,
        packetsReceived = stats.members["packetsReceived"] as? Long,
        bytesReceived = stats.members["bytesReceived"] as? BigInteger,
        framesReceived = stats.members["framesReceived"] as? Int,
        frameWidth = stats.members["frameWidth"] as? Long,
        frameHeight = stats.members["frameHeight"] as? Long,
        framesPerSecond = stats.members["framesPerSecond"] as? Double,
        framesDropped = stats.members["framesDropped"] as? Long
      )
  }
}
